### PR TITLE
Add utils/resources.py

### PR DESCRIPTION
**Pull request by Lembrun**
Concerning  #6104

(Btw there there are still some test that are localized to english only ,at some point one test says /nonexistentfile IIRC it happens fairly early in the test suite like 12%-13% ~,

I'd like to give the exact test but my fans are spinning like crazy)

### CODE REVIEW

File: qutebrowser/utils/resources.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+import contextlib
+import posixpath
+import shlex
+import mimetypes
+import pathlib
+import ctypes
+import ctypes.util
+from typing import (Any, Callable, IO, Iterator, Optional,
+                    Sequence, Tuple, Type, Union,
+                    Iterable, TypeVar, TYPE_CHECKING)
```
Comment (The-Compiler) [inline qutebrowser/utils/resources.py:41]: Many of those imports are unused - not sure why the linters don't detect that...

File: qutebrowser/utils/resources.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Resources related utilities"""
```
Comment (The-Compiler) [inline qutebrowser/utils/resources.py:20]: Nitpick:

```suggestion
"""Resources related utilities."""
```

File: qutebrowser/utils/resources.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    Args:
+        filename: The filename to open as string.
+
+    Return:
+        The file contents as a bytes object.
+    """
+    path = _resource_path(filename)
+    with _resource_keyerror_workaround():
+        return path.read_bytes()
+
```
Comment (The-Compiler) [inline qutebrowser/utils/resources.py:148]: Nitpick: Remove the blank line

File: qutebrowser/utils/resources.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        # Union[pathlib.Path, zipfile.Path] because we set "python_version = 3.6" in
+        # .mypy.ini, but the zipfiel stubs (correctly) only declare zipfile.Path with
+        # Python 3.8...
+        assert path.is_dir(), path  # type: ignore[unreachable]
+        for subpath in path.iterdir():
+            if subpath.name.endswith(ext):
+                yield posixpath.join(subdir, subpath.name)
+
+
+def preload_resources() -> None:
```
Comment (The-Compiler) [inline qutebrowser/utils/resources.py:107]: I think it'd make sense to rename this to `preload` now that it's namespaced in `resources.py`.

**Review by The-Compiler — CHANGES_REQUESTED** [review]
Thanks! Looks great overall, I added a couple of comments. Some more generic changes I'd suggest:

- Also split the resource-related tests from `tests/unit/utils/test_utils.py` into a new `test_resources.py` (so there's a rough 1:1 mapping between test files and code files).
- Add `resources.py` and the new tests file to `PERFECT_FILES` in `scripts/dev/check_coverage.py` because it has 100% test coverage.
- Rename internal functions/variables in `resources.py` to remove the `_resource_...` prefix, as it's unneeded when everything is grouped in a `resources.py` now.

**Comment by The-Compiler** [comment]
Oh, as for the failing tests - I tried reproducing with a German locale, found two failing tests and fixed them in 45387d18a514103c686be559087420199cb49995.

**Comment by Lembrun** [comment]
Yeah I was hoping the linters would spare me some more manual work but I guess it can't be helped, the test suite seems to be working on my french locale now nice !

**Comment by The-Compiler** [comment]
Thanks! Merged with some small adjustments:

- 01b57a33b Merge remote-tracking branch 'origin/pr/6234'
- 822239b12 Make private things private again

(When I mentioned dropping the `resource` prefix, I meant keeping the initial `_` which marks the functions/variables as "private".)

**Comment by Lembrun** [comment]
Yeah I know about the private marker, was just a brain fart

**Pull request merged**